### DISCUSSION TITLE
Closest point on body

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#445] [Q&A] [answered]
I might be missing something obvious here, but I'm in the process of implementing some of Godot's physics queries and one such query is `get_closest_point_to_object_volume`. Judging by [the implementation in Godot 4](https://github.com/godotengine/godot/blob/e0de3573f3fc86062763152f5a1ac62f5a986da3/servers/physics_3d/godot_space_3d.cpp#L568-L606) it seems its name is appropriate in that it gets the closest point on the body/sensor from some reference point with all its (convex) shapes considered.

I noticed you have `GJKClosestPoint::GetClosestPoints`, which seemed appropriate for this, but it relies on types that provide a `GetSupport` method, and I couldn't find any that represent a single point. Interestingly, [the Bullet implementation in Godot 3](https://github.com/godotengine/godot/blob/ca823e16e423eb6917ee4752603785795c54fd51/modules/bullet/space_bullet.cpp#L292-L343) seems to have run into this exact same problem and [worked around it](https://github.com/godotengine/godot/blob/ca823e16e423eb6917ee4752603785795c54fd51/modules/bullet/space_bullet.cpp#L302) by simply creating a sphere shape of radius 0.

Is there some other/better way of achieving this in Jolt, or am I better off replicating what they did with regards to using a zero-sized sphere?

(Neither of the two above mentioned implementations seem to care about non-convex shapes in this particular query, so GJK should be fine I guess.)

**Comment by jrouwe** [comment, marked as answer]
Hello,

The GetSupport method of a point is trivial, see [PointConvexSupport](https://github.com/jrouwe/JoltPhysics/blob/1e760859308c2b6c4f4951c40e6d80b7132aea28/Jolt/Geometry/ConvexSupport.h#L84). So I think you can implement this functionality like:

```
PointConvexSupport query_point
query_point.mPoint = the_point_to_query

shape->CollectTransformedShapes(sufficiently large bounding box, ..., collector, ...)

iterate over what collector found (or do this in the AddHit function directly)
   if (sub_shape->GetType() == EShapeType::Convex)
      convex_shape = static_cast<const ConvexShape *>(sub_shape)

      support_function = convex_shape->GetSupportFunction(IncludeConvexRadius, ...)

      GJKClosestPoint gjk
      distance = gjk.GetClosestPoints(query_point, *support_function, ..., closest_point_b)
      if (distance == 0.0f)
         closest_point = the_point_to_query // point is inside the shape, closest point is query point
      else if (distance < FLT_MAX)
         we found the closest_point_b for this sub shape, keep iterating if we find a closer one on another sub shape
```

**Reply by mihe** [reply]
Well, that's embarrassing. Not only did you have `PointConvexSupport` already, but its implementation was (as you point out) trivial to say the least.

Using `CollectTransformedShapes` was not something I had considered though. That could actually help me contain the solution into one place a lot better than the sprawling thing I came up with.

Thank you!
